core/domain/exp_services.py

```diff
@@ -1533,43 +1601,69 @@ def regenerate_exploration_and_contributors_summaries(exploration_id):
         exploration_id: str. ID of the exploration.
     """
     exploration = exp_fetchers.get_exploration_by_id(exploration_id)
-    exp_summary = _compute_summary_of_exploration(exploration)
-    exp_summary.contributors_summary = (
-        compute_exploration_contributors_summary(exp_summary.id))
-    save_exploration_summary(exp_summary)
-
-
-def _compute_summary_of_exploration(exploration):
+    exp_rights = rights_manager.get_exploration_rights(
+        exploration_id, strict=False)
+    exp_summary = exp_fetchers.get_exploration_summary_by_id(
+        exploration_id, strict=False)
+    updated_exp_summary = compute_summary_of_exploration(
+        exploration, exp_rights, exp_summary)
+    updated_exp_summary.contributors_summary = (
+        compute_exploration_contributors_summary(updated_exp_summary.id))
+    save_exploration_summary(updated_exp_summary)
+
+
+def compute_summary_of_exploration(
+    exploration,
+    exp_rights,
+    exp_summary,
+    skip_exploration_model_last_updated=False
+):
     """Create an ExplorationSummary domain object for a given Exploration
     domain object and return it.
 
     Args:
         exploration: Exploration. The exploration whose summary is to be
             computed.
+        exp_rights: ActivityRights. The exploration rights model, used
+            to compute summary.
+        exp_summary: ExplorationSummary|None. The exploration summary
```

Comment:
For this docstring, the generation in this function is for an exploration summary domain object, not a model, right? The description of the object is also wrong (it says "the exploration summary model").

That said, I don't understand why you need to pass this in -- it seems to mix things up somewhat for this function and makes it more convoluted than it was before. Instead, could this function just take in an Exploration and ActivityRights, and compute its summary from scratch? In particular:
- If you need specific stuff like contributors_summary and ratings, pass just those in (as previous_contributors_summary, previous_ratings, etc.), not the whole existing summary object. (Also, if the caller is responsible for providing these values correctly, mention that in the docstring.)
- For exploration_model_last_updated couldn't you get that from the Exploration domain object that's being passed in? (If not, pass it in as a separate arg and explain why it's not possible to use the Exploration domain object.) 
- Then you shouldn't need all the branching at the top of this function.

Reply:
Hmm, I agree with some of these.

> If you need specific stuff like contributors_summary and ratings, pass just those in (as previous_contributors_summary, previous_ratings, etc.), not the whole existing summary object. (Also, if the caller is responsible for providing these values correctly, mention that in the docstring.)

For me, it seems better to pass the whole existing summary object instead of some of its parts. Although I agree that currently the function is written a bit weirdly. 

> For exploration_model_last_updated couldn't you get that from the Exploration domain object that's being passed in? (If not, pass it in as a separate arg and explain why it's not possible to use the Exploration domain object.)

The issue is that we need to use `get_last_updated_by_human_ms` for getting the `exploration_model_last_updated`, as `get_last_updated_by_human_ms` does quite complicated fetching of models we cannot call it from Beam jobs, also we know that the last "human" update time will not be changed after running the Beam job (so we can just pass the previous value from the exp summary model).

I've added a few new functions just to compare and pick one solution:
- `compute_summary_of_exploration2` — changes according your recommendations, I don't really like the three arguments that with `previous` prefix as it seems a bit repeated
- `update_summary_of_exploration` and `generate_new_summary_of_exploration` — these two split the generation of summary for a new exploration and for an existing exploration, I like this the most as it makes it more readable in the places where we call these functions. In all places we know we work with existing explorations we will call `update_summary_of_exploration` and in all places where the summary might not yet exist we check whether it exists and then call `update_summary_of_exploration` or `generate_new_summary_of_exploration`. It introduces a slight code repetition but I think it is fine compared to better readablity.

Reply:
Thanks -- I'm fine with either approach. I think your proposal is a bit cleaner, the only thing I would suggest is to give more explanation around the "skip" argument so that a caller knows when to use it and when not to.

Also just to check, does this correctly handle the case when exps are unpublished and then published again (in terms of calculating ratings, contributors, etc)?

Reply:
Yes, in case of republishing everything is recalculated.